enhancements/cluster-scope-secret-volumes/csi-driver-host-injections.md

```diff
@@ -213,158 +163,148 @@ the underlying, associated Pod references CSI volumes using `Shares`
+
+The CSI driver will be deployed as a DaemonSet so that shared Secrets/ConfigMaps can be added to pods running on the node.
+Pods for this DaemonSet will need to run as privileged to access host paths on the node.
+The CSI driver will only need to support "ephemeral" volumes that are specified by the `csi:` volume type on Pods.
+There is no requirement for this CSI driver to provision persistent volumes.
 
 Some details around inline ephemeral volumes:
 - The employ a subset of the CSI specification flows, specifically only the `NodePublish` and `NodeUnpublish` CSI steps
 - the underlying k8s api object `CSIVolumeSource` has the same `volumeAttributes` for storing metadata to be passed to the CSI driver
 - there is only a single `LocalObjectReference` for a `Secret` reference in the same namespace as the consuming pod.  So that aspect does not help us, and will not factor into the solution.
 - The `volumeAttributes` field for metadata gives us enough capabilities for expressing both the name, namespace, and type of sharable API object to be consumed.  And who is requesting access.
 
+#### Enhancements to Security Context Constraints
+
+OpenShift uses Security Context Constraints (SCCs) to regulate the capabilities granted to Pods, based on the user creating the pod and the service account assigned to the pod.
+Amongst other things, SCCs determine if a pod can mount a specific volume type.
+SCCs deny mounting a volume unless its type is explicitly allowed.
+The most restrictive SCC only allows volume types that are known to be isolated from the host filesystem (`secret`, `configMap`, `persistentVolumeClaim`, etc.).
+
+SCCs as designed today have an "all or nothing" approach when it comes securing CSI volume mounts.
+This is not sufficient as cluster admins are free to install third party CSI drivers that support `csi` volume mounts.
+Because we want to deliver this CSI driver in the payload, we need ensure that this CSI driver is usable while ensuring third party CSI drivers can be restricted.
+
+The following API extends SCCs to restrict the allowed CSI drivers used for ephemeral CSI volumes, in the same manner that Flexvolumes are restricted:
+
+```go
+type SecurityContextConstraints struct {
+  // existing API
+  ...
+  //
+
+  Volumes []FSType
+
+  // AllowedCSIVolumes is an allowlist of permitted ephemeral CSI volumes.
+  // Empty or nil indicates that all CSI drivers may be used.
+  // This parameter is effective only when the usage of CSI volumes is allowed in the "Volumes" field.
+	// +optional
+  AllowedCSIVolumes []AllowedCSIVolume
+}
+
+// AllowedCSIVolume represents a single CSI volume driver that is allowed to be used.
+type AllowedCSIVolume struct {
+  // Driver is the name of the CSI driver that supports ephemeral CSI volumes.
+  Driver string
+}
+```
+
+With this API, the default SCCs delivered via OpenShift can be updated to include the projected resource CSI driver as an allowed driver.
+
```

Comment:
@deads2k per our discussion on Slack, I've added a section here to extend SCCs to support these CSI drivers. I borrowed the API used to restrict flex volumes - is this still a valid approach?